src/Bicep.Core/TypeSystem/Radius/V3/KnownComponents.cs

```diff
@@ -653,19 +653,92 @@ public static ComponentData MakeMongoDB()
+                properties: Array.Empty<TypeProperty>(),
+                additionalPropertiesType: keysItemType,
+                additionalPropertiesFlags: TypePropertyFlags.None,
+                functions: null);
+
+            var certificatesItemType = new ObjectType(
+                name: "certificates",
+                validationFlags: TypeSymbolValidationFlags.Default,
+                properties: new TypeProperty[] {
+                    new TypeProperty("name", LanguageConstants.String, TypePropertyFlags.Required, "Name of the Azure KeyVault certificate"),
+                    new TypeProperty("version", LanguageConstants.String, TypePropertyFlags.None, "Version of the certificate. Latest by default"),
+                    new TypeProperty("encoding", encodingType, TypePropertyFlags.None, "Encoding format of the certificate"),
+                    new TypeProperty("format", formatType, TypePropertyFlags.None, "Format of the certificate"),
+                    new TypeProperty("value", valueType, TypePropertyFlags.Required,"Value to be downloaded from the Azure KeyVault"),
+                    },
+                additionalPropertiesType: null,
+                additionalPropertiesFlags: TypePropertyFlags.None,
+                functions: null);
+
+            var certificates = new ObjectType(
+                name: "certificates",
+                validationFlags: TypeSymbolValidationFlags.Default,
+                properties: Array.Empty<TypeProperty>(),
+                additionalPropertiesType: certificatesItemType,
+                additionalPropertiesFlags: TypePropertyFlags.None,
+                functions: null);
+
+
+            var persistentVolumeKindType = new UnionType(
+                "persistent volume kind",
+                ImmutableArray.Create<ITypeReference>(
+                    new StringLiteralType("azure.com.fileshare"),
+                    new StringLiteralType("azure.com.keyvault")
+                )
+            );
 
             return new ComponentData()
             {
-                Type = new ThreePartType(null, "Volume", ""),
+                Type = new ThreePartType(null, "Volume", RadiusResources.CategoryComponent),
                 Properties =
                 {
-                    new TypeProperty("kind", configKindType, TypePropertyFlags.Required),
-                    new TypeProperty("managed", LanguageConstants.Bool, TypePropertyFlags.None),
-                    new TypeProperty("resource", LanguageConstants.String, TypePropertyFlags.None),
-                },
+                    new TypeProperty("managed", LanguageConstants.Bool, TypePropertyFlags.None, "Managed by Radius"),
+                    new TypeProperty("resource", LanguageConstants.String, TypePropertyFlags.None, "Resource ID for user managed resource"),
+                    new TypeProperty("kind", persistentVolumeKindType, TypePropertyFlags.Required, "Persistent volume kind"),
```

Comment:
Ideally, we want the keyvault related properties to show up only when the kind = azure.com.keyvault. With the current design, this looked possible only if I add another nesting level within the volume component which is not the schema we want. Looking for ideas to implement this.

Reply:
I think you would implement this the same way we implement traits: https://github.com/rynowak/bicep/blob/radius-compiler/src/Bicep.Core/TypeSystem/Radius/V3/CommonTraits.cs#L58

Reply:
I had tried a similar approach but that results in an extra level in between which we do not want in the schema. For example:-
resource afs 'VolumeComponent' = {
    name: 'myshare2'
    properties:{
      foo:{
        kind:'azure.com.keyvault'
        certificates:{
          ...
        }
      }
    }
  }

I do not want the typeproperty foo in the component object but want a different set directly under properties depending on the 'kind'

Reply:
I see, you may be right about this then.

I'd say we leave this as is as in other places where we have a kind, we don't have a discriminated union check as well.